src/NLog/LogFactory.cs

```diff
@@ -580,9 +583,103 @@ public void Flush(AsyncContinuation asyncContinuation, TimeSpan timeout)
+        /// <param name="asyncTimeout">Optional timeout that guarantees that completed notication is called.</param>
+        /// <returns></returns>
+        private static AsyncContinuation FlushAllTargetsAsync(LoggingConfiguration loggingConfiguration, AsyncContinuation asyncContinuation, TimeSpan? asyncTimeout)
+        {
+            var targets = loggingConfiguration.GetAllTargetsToFlush();
+            var pendingTargets = new HashSet<Target>(targets, SingleItemOptimizedHashSet<Target>.ReferenceEqualityComparer.Default);
+            AsynchronousAction<Target> flushAction = (target, cont) =>
+            {
+                target.Flush(ex =>
+                {
+                    if (ex != null)
+                        InternalLogger.Warn(ex, "Flush failed for target {0}(Name={1})", target.GetType(), target.Name);
+                    lock (pendingTargets)
+                    {
+                        pendingTargets.Remove(target);
+                    }
+                    cont(ex);
+                });
+            };
+            AsyncContinuation flushContinuation = (ex) =>
+            {
+                lock (pendingTargets)
+                {
+                    foreach (var pendingTarget in pendingTargets)
+                        InternalLogger.Debug("Flush timeout for target {0}(Name={1})", pendingTarget.GetType(), pendingTarget.Name);
+                    pendingTargets.Clear();
+                }
+                if (ex != null)
+                    InternalLogger.Warn(ex, "Flush completed with errors");
+                else
+                    InternalLogger.Debug("Flush completed");
+                asyncContinuation(ex);
+            };
+            if (asyncTimeout.HasValue)
+            {
+                flushContinuation = AsyncHelpers.WithTimeout(flushContinuation, asyncTimeout.Value);
+            }
+            else
+            {
+                flushContinuation = AsyncHelpers.PreventMultipleCalls(flushContinuation);
+            }
+            InternalLogger.Trace("Flushing all {0} targets...", targets.Count);
+            AsyncHelpers.ForEachItemInParallel(targets, flushContinuation, flushAction);
+            return flushContinuation;
+        }
+
+        private static bool FlushAllTargetsSync(LoggingConfiguration oldConfig, TimeSpan timeout, bool throwExceptions)
+        {
+            Exception lastException = null;
+
+            try
+            {
+                ManualResetEvent flushCompletedEvent = new ManualResetEvent(false);
+
+                var flushContinuation = FlushAllTargetsAsync(oldConfig, (ex) =>
+                {
+                    if (ex != null)
+                        lastException = ex;
+                    flushCompletedEvent.Set();
+                }, null);
+
+                bool flushCompleted = flushCompletedEvent.WaitOne(timeout);
+                if (!flushCompleted)
+                    flushContinuation(new TimeoutException("Timeout."));
```

Comment:
proposal for the TimeoutException message:

"Timeout when flushing all targets, timeout used: X sec"

Reply:
Fine with me. Was just trying to ensure that the user would not see the impact of the change, unless using the InternalLogger.

Reply:
But it will not change the `TimeoutException` returned by `TimeoutContinuation`. But I think most users will be calling the sync-`Flush()`.